<a href="https://colab.research.google.com/github/GabrielKonno/GabrielKonno/blob/main/Modeling_Pr%C3%A1tica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from sklearn import datasets
import pandas as pd

In [3]:
housing = datasets.fetch_california_housing() #importando dataset

In [4]:
print(housing)

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
          37.88      , -122.23      ],
       [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
          37.86      , -122.22      ],
       [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
          37.85      , -122.24      ],
       ...,
       [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
          39.43      , -121.22      ],
       [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
          39.43      , -121.32      ],
       [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
          39.37      , -121.24      ]]), 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]), 'frame': None, 'target_names': ['MedHouseVal'], 'feature_names': ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude'], 'DESCR': '.. _california_housing_dataset:\n\nCalifornia Housing dataset\n-

In [5]:
df = pd.DataFrame(housing.data, columns=housing.feature_names) #Transformando em DF

In [6]:
housing.keys() #Vendo as keys

dict_keys(['data', 'target', 'frame', 'target_names', 'feature_names', 'DESCR'])

In [7]:
print(housing.DESCR)

.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

:Number of Instances: 20640

:Number of Attributes: 8 numeric, predictive attributes and the target

:Attribute Information:
    - MedInc        median income in block group
    - HouseAge      median house age in block group
    - AveRooms      average number of rooms per household
    - AveBedrms     average number of bedrooms per household
    - Population    block group population
    - AveOccup      average number of household members
    - Latitude      block group latitude
    - Longitude     block group longitude

:Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,000).

This dataset was derived from the 1990 U.S. census, using one row per ce

In [8]:
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [9]:
print(housing.target)

[4.526 3.585 3.521 ... 0.923 0.847 0.894]


In [10]:
df['MedValue'] = housing.target

In [11]:
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  MedValue  
0    -122.23     4.526  
1    -122.22     3.585  
2    -122.24     3.521  
3    -122.25     3.413  
4    -122.25     3.422

In [12]:
x = housing.data # Características Casas
y = housing.target # Preço das Casas


# Modeling Techniques
1. Regressão Linear SKlearn <https://scikit-learn.org/dev/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression>

2. SVR <https://scikit-learn.org/dev/modules/generated/sklearn.svm.SVR.html#sklearn.svm.SVR>

3. Decision Tree Regression - XGBoost <https://xgboost.readthedocs.io/en/stable/tutorials/index.html>


## Modeling Assumptions:

Apenas variáveis numéricas

## Métricas de Avaliação do Modelo:
### Para penalizar erros grandes:
1. MSE
2. RMSE
<https://scikit-learn.org/dev/api/sklearn.metrics.html>

# Teste Design
## Dataset split:

Separação Train-Test de 20% para teste por método SKlearn

In [13]:
# Técnica 1. Regressão Linear

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression
import numpy as np

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

regLin = LinearRegression()

regLin = regLin.fit(x_train, y_train)

yPredLin = regLin.predict(x_test)

MSE = mean_squared_error(y_test, yPredLin)
RMSE = np.sqrt(MSE)

print("LINEAR MSE:", MSE)
print("LINEAR RMSE:", RMSE)

LINEAR MSE: 0.5558915986952422
LINEAR RMSE: 0.7455813830127749


In [14]:
# Técnica 2. SVR

from sklearn.svm import SVR

regSVR = SVR().fit(x_train, y_train)

yPredSVR = regSVR.predict(x_test)

MSESVR = mean_squared_error(y_test, yPredSVR)
RMSESVR = np.sqrt(MSESVR)

print("SVR MSESVR:", MSESVR)
print("SVR RMSE:", RMSESVR)

SVR MSESVR: 1.3320115421348744
SVR RMSE: 1.1541280440812771


In [15]:
# Técnica 3. Decision Tree Regression (XGBoost)

from xgboost import XGBRegressor

regXGB = XGBRegressor().fit(x_train, y_train)

yPredXGB = regXGB.predict(x_test)

MSEXGB = mean_squared_error(y_test, yPredXGB)
RMSEXGB = np.sqrt(MSEXGB)

print("XGB MSE:", MSEXGB)
print("XGB RMSE:", RMSEXGB)

XGB MSE: 0.2225899267544737
XGB RMSE: 0.4717943691423984


# Interpretando os resultados:

1. Modelo Linear (Regressão Linear):
MSE: 0.5559
RMSE: 0.7456
A regressão linear tem um MSE de 0.5559, o que indica que o modelo está errando, em média, aproximadamente $74.560 (SQR(0.5559) × 100.000) ao prever os valores das casas. Esse erro não é insignificante, mas pode ser aceitável dependendo do contexto do mercado de imóveis e dos valores em questão.

2. Modelo SVR (Support Vector Regressor):
MSE: 1.3320
RMSE: 1.1541
O modelo SVR apresentou um MSE de 1.3320, indicando um desempenho significativamente pior em relação ao modelo linear. O RMSE de 1.1541 mostra que o erro médio nas previsões é de $115.410 (SQR(1.3320)×100.000), o que pode ser considerado um erro muito alto para o seu objetivo. O SVR claramente não se ajustou bem aos dados, sugerindo que ele não é a melhor escolha para esse problema em específico.

3. Modelo XGBoost (Árvore de Decisão de Regressão):
MSE: 0.2226
RMSE: 0.4718
O XGBoost foi o que teve o melhor desempenho entre os três modelos, com um MSE de 0.2226 e um RMSE de 0.4718. Isso indica que, em média, o erro nas previsões é de $47.180 (0.2226×100.000), um valor consideravelmente menor em comparação aos outros modelos. Este é o modelo mais preciso e ajustado aos dados, de acordo com essa métrica.

##Conclusão e Comparação:

### XGBoost:
Este modelo apresentou o menor MSE e RMSE, o que sugere que ele oferece as previsões mais precisas entre os três. O erro médio de $47.180 é substancialmente menor do que os erros nos outros dois modelos.

### Regressão Linear:
Embora o modelo linear tenha apresentado resultados razoáveis, com um erro médio de $74.560, ele ainda é inferior ao XGBoost. Pode ser uma opção viável se você busca simplicidade no modelo, mas não oferece a mesma precisão.

### SVR:
O SVR teve um desempenho consideravelmente pior, com um erro médio de $115.410. Este modelo parece inadequado para esse problema, pelo menos com os parâmetros testados.

#Otimização de Hiperparâmetros

Utilizando GridSearchCV do SKlearn

<https://scikit-learn.org/dev/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV>


In [16]:
from sklearn.model_selection import GridSearchCV

In [17]:
regXGB.get_params().keys()

dict_keys(['objective', 'base_score', 'booster', 'callbacks', 'colsample_bylevel', 'colsample_bynode', 'colsample_bytree', 'device', 'early_stopping_rounds', 'enable_categorical', 'eval_metric', 'feature_types', 'gamma', 'grow_policy', 'importance_type', 'interaction_constraints', 'learning_rate', 'max_bin', 'max_cat_threshold', 'max_cat_to_onehot', 'max_delta_step', 'max_depth', 'max_leaves', 'min_child_weight', 'missing', 'monotone_constraints', 'multi_strategy', 'n_estimators', 'n_jobs', 'num_parallel_tree', 'random_state', 'reg_alpha', 'reg_lambda', 'sampling_method', 'scale_pos_weight', 'subsample', 'tree_method', 'validate_parameters', 'verbosity'])

In [18]:
params = {
    'max_depth':[5, 6, 7],
    'learning_rate': [0.1, 0.2, 0.3],
    'objective' : ['reg:squarederror'],
    'booster' : ['gbtree'],
    'n_jobs' : [5],
    'gamma' : [0, 1],
    'min_child_weight' : [1, 3],
    'max_delta_step' : [0, 1],
    'subsample' : [0.5, 1]
}

In [21]:
xgbGrid = GridSearchCV(XGBRegressor(), params, refit = 'neg_mean_square_error', verbose = True) # Verbose narra o que tá acontecendo

In [22]:
xgbGridModel = xgbGrid.fit(x_train, y_train)

Fitting 5 folds for each of 144 candidates, totalling 720 fits


/usr/local/lib/python3.10/dist-packages/numpy/ma/core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


In [23]:
xgbGridModel.best_params_ # best_params_ lista a melhor combinação de parâmetros

{'booster': 'gbtree',
 'gamma': 0,
 'learning_rate': 0.2,
 'max_delta_step': 1,
 'max_depth': 7,
 'min_child_weight': 1,
 'n_jobs': 5,
 'objective': 'reg:squarederror',
 'subsample': 1}

In [24]:
yGrid = xgbGridModel.predict(x_test)

MSEGrid = mean_squared_error(y_test, yGrid)
RMSEGrid = np.sqrt(MSEGrid)

print("XGB Grid MSE:", MSEGrid)
print("XGB Grid RMSE:", RMSEGrid)

XGB Grid MSE: 0.21638147438949515
XGB Grid RMSE: 0.4651682216032122


# Conclusão

Temos, ao fim, o MSE e RMSE com os melhores parâmetros encontrados pelo GridSearch entre as variações dadas.
